### ISSUE TITLE
[java] No Method Calls in Method Parameters

### ISSUE THREAD

**Issue by numeralnathan** [pmd/pmd#2844]
**Proposed Rule Name:**

MethodCallInMethodParameters

**Proposed Category:** Code Style

**Description:**

Please create a rule that flags code that has method calls inside the parameter list for a method call.

**Code Sample:** 

Code to be flagged because it is harder to read.  This is a tame example.  There are much more difficult to read examples in the wild.
```
listOfPaths.addAll(extract(file.getAbsolutePath(), file.getParent(), null));
```

"Correct" code because it is easier to read.  The assignments to local variables could make the code self-documenting.
```
absolutePath  = file.getAbsolutePath();
parentFile    = file.getParent();
extractedFile = extract(absolutePath, parentFile, null);

listOfPaths.addAll(extractedFile);
```

**Possible Properties:**

Unknown

**Comment by Fernal73** [comment]
An exception must be made for test classes when asserts are invoked on method calls.

**Comment by numeralnathan** [comment]
@Fernal73 I do not think asserts should be an exception.  Consider these two methods that test the same thing.  The `extractHardToRead()` would take me a while to figure out which parameter is what.  The `extractEasyToRead()` makes it very clear very fast.

```
@Test
public static void extractEasyToRead()
{
   String path, expect;
   File source, parent, actual, expect;

   expect = getFileName("extractTest");
   source = new File(expect);
   source = source.getAbsoluteFile();
   path   = source.getAbsolutePath();
   parent = source.getParent();
   actual = extract(absolutePath, parentFile, null);

   Assert.assertEquals(actual, expect);
}

@Test
public static void extractHardToRead()
{
   File source;

   source = new File("test").getAbsoluteFile();

   Assert.assertEquals(extract(source.getAbsolutePath(), source.getParent(), null), getFileName("extractTest"));
}
```

I suppose another way to look at this rule is to keep each line focused on 1 thing just like methods and classes should have a single focus.

**Comment by Fernal73** [comment]
> @Fernal73 I do not think asserts should be an exception. Consider these two methods that test the same thing. The `extractHardToRead()` would take me a while to figure out which parameter is what. The `extractEasyToRead()` makes it very clear very fast.
> 
> ```
> @Test
> public static void extractEasyToRead()
> {
>    String path, expect;
>    File source, parent, actual, expect;
> 
>    expect = getFileName("extractTest");
>    source = new File(expect);
>    source = source.getAbsoluteFile();
>    path   = source.getAbsolutePath();
>    parent = source.getParent();
>    actual = extract(absolutePath, parentFile, null);
> 
>    Assert.assertEquals(actual, expect);
> }
> 
> @Test
> public static void extractHardToRead()
> {
>    File source;
> 
>    source = new File("test").getAbsoluteFile();
> 
>    Assert.assertEquals(extract(source.getAbsolutePath(), source.getParent(), null), getFileName("extractTest"));
> }
> ```
> 
> I suppose another way to look at this rule is to keep each line focused on 1 thing just like methods and classes should have a single focus.
```
assertEquals(expect, extract(absolutePath, parent, null), MESSAGE);
```
looks quite different.

Similarly,
```java
assertTrue(isValueTrue(param)), MESSAGE);
```
is quite compact.

One level of nesting can be allowed and a restriction on number of parameters in a nested function call. 

This doesn't necessarily have to apply just to test classes, though.

**Comment by Fernal73** [comment]
> 
> ```
> assertEquals(expect, extract(absolutePath, parent, null), MESSAGE);
> ```
> 
> looks quite different.
> 
> Similarly,
> 
> ```java
> assertTrue(isValueTrue(param)), MESSAGE);
> ```
> 
> is quite compact.
> 
> One level of nesting can be allowed and a restriction on number of parameters in a nested function call.
> 
> This doesn't necessarily have to apply just to test classes, though.

I suppose you could actually have another rule that permits only **actual** parameters , the values being tested, to have method calls, the other parameters have to be variables.

**Comment by numeralnathan** [comment]
@Fernal73 I added a parameter to the original post to allow for a certain number of levels of methods in parameter lists.

**Comment by Fernal73** [comment]
> @Fernal73 I added a parameter to the original post to allow for a certain number of levels of methods in parameter lists.

I'd go with allowing only one outer method parameter to have a nested call as default when nestDepth = 1. Anything more looks crowded. 

Or should we restrict it to only parameterless or one parameter methods? 

When it comes to code style, the aesthetics become equally important as well!!!

**Comment by numeralnathan** [comment]
> Or should we restrict it to only parameterless or one parameter methods?

I added `nestedMethodParameterLimit` to the original post.  This could cover your idea.  Can you think of a better rule parameter name?

If one were to set the two rule parameters to Integer.MAX_VALUE, then this rule basically becomes inert.  With these two parameters, hopefully it provides enough flexibility to meet a variety of aesthetics.

**Comment by Fernal73** [comment]
Excellent!
That should allow users to add the rule to their ruleset without having to retrospectively correct the violations in their existing code base. Immediately, that is!
An auto-correction facility would be easiest implemented for the default values.

**Comment by Fernal73** [comment]
> @Fernal73 I do not think asserts should be an exception. 

Do you think this rule could be extended to disallow method calls in assert statements?

If method calls are allowed, the method definition must be annotated with Checker framework's SideEffectFree or `Pure` annotations.

**Comment by Fernal73** [comment]
https://github.com/Lanchon/r8/blob/e22a77d26e30615bfc9b5f598bb0f58f10dd082c/src/test/java/com/android/tools/r8/shaking/ifrule/IfOnAnnotationTestClasses.java
